### DISCUSSION TITLE
Getting PointIO devices?

### DISCUSSION THREAD

**Discussion by 13igred** [ottowayi/pycomm3#283] [Q&A] [answered]
Hello,

Is it possible to get the module information of PointIO devices?

![image](https://github.com/ottowayi/pycomm3/assets/81141479/0f9dc35e-26d9-42a2-8abe-edc6c011d78c)

I have several PLCs with a lot of PointIO and I want to read for, example the revision number, part number, ect. Basically the same information you get when you do a get_module_info.

**Comment by ottowayi** [comment, marked as answer]
Yeah, unfortunately that method isn't very sophisticated and doesn't work for this.  I plan on replacing it and adding a bunch more features to the `CIPDriver`, but until then you can do it but have to do it manually.  Here is a small example that works, you can probably modify it to your needs:

``` 
from pycomm3 import CIPDriver, ModuleIdentityObject, ClassCode, Services

entrypoint = '10.1.128.45'
pointio_route = '1/1/2/192.168.1.41'

with CIPDriver(entrypoint) as plc:
    
    for slot in range(8):
        resp = plc.generic_message(
            service=Services.get_attributes_all,
            class_code=ClassCode.identity_object,
            instance=1,
            data_type=ModuleIdentityObject,
            route_path=f'{pointio_route}/1/{slot}',
            connected=False,
            unconnected_send=True,
        )
    
        print(f'{slot}: {resp.value["product_name"] if resp else "EMPTY"}'))
```
![image](https://github.com/ottowayi/pycomm3/assets/5274950/5ffcf8ca-cfa3-4368-9686-34b569097298)

This is for a rack that's behind a controllogix chassis, if you can access your point io rack directly by IP, you should be able to omit the `pointio_route` and make your `route_path` just `1/{slot}`.

**Reply by 13igred** [reply]
Works like a charm, thanks and sorry for the late reply. I did not see the email.
